packages/dds/tree/src/util/rangeMap.ts

```diff
@@ -6,229 +6,239 @@
 import { oob } from "@fluidframework/core-utils/internal";
 
 /**
- * A map keyed on integers allowing reading and writing contiguous ranges of integer keys.
- *
- * TODO: We should avoid the direct exposure of RangeEntry. AB#7414
+ * RangeMap represents a mapping from integers to values of type T or undefined.
+ * The values for a range of consecutive keys can be changed or queried in a single operation.
  */
-export type RangeMap<T> = RangeEntry<T>[];
+export class RangeMap<T> {
+	private readonly entries: RangeEntry<T>[];
 
-export interface RangeEntry<T> {
-	start: number;
-	length: number;
-	value: T;
-}
+	public constructor(initialEntries?: RangeEntry<T>[]) {
+		this.entries = initialEntries ? [...initialEntries] : [];
+	}
 
-/**
- * The result of a query about a range of keys.
- */
-export interface RangeQueryResult<T> {
 	/**
-	 * The value of the first key in the query range.
+	 * Retrieves all entries from the rangeMap.
+	 * @returns An array of RangeEntryResult objects, each containing the start index, length, and value of a contiguous range.
 	 */
-	value: T | undefined;
+	public getAllEntries(): RangeQueryResult<T>[] {
```

Comment:
This will allow the caller to mutate the guts of the RangeMap. Making the return type `readonly RangeQueryResult<T>[]` will help a bit, but the individual `RangeQueryResult`s are mutable.
Unless we need them to be mutable, I'd be fine with making the fields on `RangeQueryResult` immutable. API-wise, it's typically best to return a mutable copy of things, but I'm not sure it makes sense to pay the copy cost here.